### ISSUE TITLE
Applying transformations to the hidden variables

### ISSUE THREAD

**Issue by wil-j-wil** [ICL-SML/Doubly-Stochastic-DGP#26]
Hi Hugh,

Not sure if this is the best place to ask, but is it possible with your implementation to apply some fixed transformations to the hidden variables in the Deep GP model? Since you are using the noise-free case p(f_i)=p(h_i), there's no explicit representation of the hidden variables. So if I want to apply a transformation, e.g. p(h_i) = p(exp(f_i), sig_i^2), it's not straightforward.

I'm able to apply the transformation to the MC samples as they pass through the model when calculating the likelihood expectation, but when doing so the hyperparameters no longer tune properly, and actually diverge. Can you spot why this might be the case / what changes are needed to allow for transformations of the hidden variables?

Regards,

Will

**Comment by hughsalimbeni** [comment]
Yes, you can certainly do this, and indeed I have done this myself for [non-stationary models](http://bayesiandeeplearning.org/2017/papers/49.pdf). One way to implement a transform between layers is to put the transformation on the kernel. So e.g. if you want to put an `exp` between layer1 and layer2 in, you could do:

```python
class f_RBF(gpflow.kernels.RBF):
    # applies the transformation f to the inputs
    def __init__(self, f, *args, **kw):
        gpflow.kernels.RBF.__init__(self, *args, **kw)
        self.f = f
        
    def K(self, X, X2=None):
        X = self.f(X)
        if X2 is not None:
            X2 = self.f(X2)
        return gpflow.kernels.RBF(X, X2)
    
    def Kdiag(self, X):
        return gpflow.kernels.RBF(self.f(X)) # not necessary for stationary kernels, but we can do this anyway

kern1 = RBF(...)
kern2 = f_RBF(tf.exp, ...)
```
NB, the values you'll get from `predict_all_layers` are not transformed, so you'll have to add the exp manually if you want plots.

If you want a trainable parameter in the transform, then `f` would need to be an `Parameterized` object that implements `__call__`

If you want the transformation on the very final layer (i.e. just before the likelihood) then this less easy as the likelihood variational expectations will break. The only way to do this would be to write a custom likelihood and use quadrature for the variational expectations (or else use a sampling method).

Regarding hyperparameter stability, I'm not sure without seeing a specific example, but I'd be wary of `exp` without some thought to initialization as the variance might get very large and you might need a more cautious learning rate (or more samples to lower the variance). I usually have gone for `softplus` in the past if I just want to make things positive.

**Comment by wil-j-wil** [comment]
Great, thanks Hugh, that's really useful. That approach (transforming the input in the kernel) does seem to be more stable than transforming the output of the previous GP - although I haven't quite figured out how the two approaches are different yet.
